rotkehlchen/chain/ethereum/modules/uniswap/v2/decoder.py

```diff
@@ -0,0 +1,105 @@
+            out_event = in_event = None
+
+            # amount_in is the amount that enters the pool and amount_out the one
+            # that leaves the pool
+            amount_in_0, amount_in_1 = hex_or_bytes_to_int(tx_log.data[0:32]), hex_or_bytes_to_int(tx_log.data[32:64])  # noqa: E501
+            amount_out_0, amount_out_1 = hex_or_bytes_to_int(tx_log.data[64:96]), hex_or_bytes_to_int(tx_log.data[96:128])  # noqa: E501
+            amount_in, amount_out = amount_in_0, amount_out_0
+            if amount_in == ZERO:
+                amount_in = amount_in_1
+            if amount_out == ZERO:
+                amount_out = amount_out_1
+
+            received_eth = ZERO
+            for event in decoded_events:
+                if event.asset == A_ETH and event.event_type == HistoryEventType.RECEIVE:
+                    received_eth += event.balance.amount
+
+            for event in decoded_events:
+                # When we look for the spend event we have to take into consideration the case
+                # where not all the ETH is converted. The ETH that is not converted is returned
+                # in an internal transaction to the user.
+                if (
+                    event.event_type == HistoryEventType.SPEND and
+                    (
+                        event.balance.amount == (spend_eth := asset_normalized_value(amount_in, event.asset)) or  # noqa: E501
+                        event.asset == A_ETH and spend_eth + received_eth == event.balance.amount
+                    )
+                ):
+                    event.event_type = HistoryEventType.TRADE
+                    event.event_subtype = HistoryEventSubType.SPEND
+                    event.counterparty = CPT_UNISWAP_V2
+                    event.notes = f'Swap {event.balance.amount} {event.asset.symbol} in uniswap-v2 from {event.location_label}'  # noqa: E501
+                    out_event = event
+                elif (
+                    (maybe_buyer == transaction.from_address or event.asset == A_ETH) and
+                    event.event_type == HistoryEventType.RECEIVE and
+                    event.balance.amount == asset_normalized_value(amount_out, event.asset)
+                ):
+                    event.event_type = HistoryEventType.TRADE
+                    event.event_subtype = HistoryEventSubType.RECEIVE
+                    event.counterparty = CPT_UNISWAP_V2
+                    event.notes = f'Receive {event.balance.amount} {event.asset.symbol} in uniswap-v2 from {event.location_label}'  # noqa: E501
+                    in_event = event
+                elif (
+                    event.event_type == HistoryEventType.RECEIVE and
+                    event.balance.amount != asset_normalized_value(amount_out, event.asset)
+                ):
+                    # Those are assets returned due to a change in the swap price
+                    event.event_type = HistoryEventType.TRANSFER
```

Comment:
Just some thoughts. We are iterating here through all the events and adjusting them. Is it possible that we modify by an accident, let's say, `1inch`'s event if it will have the same amounts, for example? Or is it impossible?

Reply:
If it has the same exact amounts and sender/receiver and is part of the same transaction then it is indeed possible to touch an unrelated event.

In case that happens though out in the wild then that would be a really really weird edge case and can't really think of a good way to handle it.

We would need to see an example transaction to see such a case and how to handle it.

Reply:
1inch events are always the latest in the list of events so will get decoded the last and the moment you change the history event type from `SPEND/RECEIVE` they don't get decoded by the others decoders in those trades